server/src/main/java/io/spine/server/delivery/Delivery.java

```diff
@@ -475,4 +518,60 @@ protected void onShardUpdated(InboxMessage message) {
                        .collect(groupingBy(m -> m.getInboxId()
                                                  .getTypeUrl()));
     }
+
+    /**
+     * Listens to the signals dispatched via the {@code MulticastBus}es and notifies the shard
+     * observers of a new {@code InboxMessage} only when it has been dispatched to all of
+     * its multicast targets.
+     */
+    private class DeliveryDispatchListener implements MulticastDispatchListener {
+
+        private final Multimap<SignalId, InboxMessage> pending =
+                synchronizedListMultimap(MultimapBuilder.hashKeys()
+                                                        .arrayListValues()
+                                                        .build());
+
+
+        private final Set<SignalId> currentlyDispatching = synchronizedSet(new HashSet<>());
```

Comment:
As discussed vocally, we use a `Synchronized` instead of a `Concurrent` since the concurrent-one allows simultaneous access. Meaning that writes may happen at the very same time as reads, and write results may yet not be visible. What we want is to make sure that everything written is visible for reading.